Improve NodePublishVolume and NodeStageVolume

okartau opened the pull request:
NodePublishVolume performs checks for attributes compatibility if volume already mounted.
If StagingTargetPath is not set, return FAILED_PRECONDITION.

Resolves: #462 
Resolves: #465

okartau commented:
I just realized that if Capability/mount_flags are present in NodePublishVolume request, current code never uses those in mounting.
I think in reality mount options are not set in request, at least we have not seen such yet (?)
The current PR contains log message to trace given mount option, so we will at least see and log what is given in request.
If we start comparing those mount flags, it means we also need to start to use those in mount.
The spec says these are "The mount options that can be used for the volume",
does the "can be used" actually mean "must be used" ?

okartau commented:
Removed test-repeater part which was  present only for testing, and WIP.

avalluri commented on pkg/pmem-csi-driver/nodeserver.go:
```diff
@@ -343,7 +346,6 @@ func (ns *nodeServer) NodeUnstageVolume(ctx context.Context, req *csi.NodeUnstag
 	defer volumeMutex.UnlockKey(req.GetVolumeId())
 
-	// showing for debug:
 	klog.V(4).Infof("NodeUnStageVolume: VolumeID:%v Staging target path:%v",
```
In this case, we can remove this logging?

okartau replied:
My idea was to have log msg in all handling functions near start, so that we see the start of handling. There is log message later, but in the middle there may be return with another msg. So my reasoning has been, lets start each handler with message. Level is 4 so we see this only in debug and test cases. I wanted to remove comment above log line because this does not add more value, level 4 of logging shows already it's for dev. purposes.

avalluri replied:
> My idea was to have log msg in all handling functions near start, so that we see the start of handling.

I fully agree with your point. My point was we can remove the existing log message in line 349, which is a duplicate of this newly added logging line. Or I missing something?

avalluri replied:
Opps sorry, both are diffrent functions. Nevermind

avalluri commented on pkg/pmem-csi-driver/nodeserver.go:
```diff
@@ -42,7 +43,12 @@ type nodeServer struct {
 	// Driver deployed to provision only ephemeral volumes(only for Kubernetes v1.15)
-	mounter mount.Interface
+	mounter       mount.Interface
+	volReadOnly   map[string]bool
```
How about introducing a new structure and keeping all these new params as fields in it. Then we can maintain a single `VolumeId:Params` map, easy to maintain. If I remember correctly we used to have such type earlier but removed as found no real need that time.

okartau replied:
yes, good idea. I started with that, but I realized now the creation and deletion happen in two places, so we can not just conveniently get rid of whole map entry in one place.
Another related question arises: I added map for fsType, which is there now for completeness, but is not actually used. This part is created and used in different function.
Can we perhaps leave fsType out, the we can have one map handling in single place.

avalluri replied:
> creation and deletion happen in two places, so we can not just conveniently get rid of whole map entry in one place.

Yes, If I understood correctly the change, the two possible places depend on type of volume: ephemeral or persistent.  So for persistent volumes the point of creation/deletion of map entry in Stage/Unstage and for ephemeral its publish/unpublish.

avalluri commented on pkg/pmem-csi-driver/nodeserver.go:
```diff
@@ -108,11 +118,15 @@ func (ns *nodeServer) NodePublishVolume(ctx context.Context, req *csi.NodePublis
+	mountFlags := req.GetVolumeCapability().GetMount().GetMountFlags()
+	readOnly := req.GetReadonly()
+	fsType := req.GetVolumeCapability().GetMount().GetFsType()
+	klog.V(3).Infof("NodePublishVolume request: VolID:%s targetpath:%v sourcepath:%v readonly:%v Capab/mountflags:%v Capab/fsType:%v",
```
Can you change `Capab/mountflags` to `mount flags:` in the log message. Capab is not informative.

okartau replied:
Changed in current code, just pushed

avalluri commented on pkg/pmem-csi-driver/nodeserver.go:
```diff
@@ -209,14 +236,21 @@ func (ns *nodeServer) NodePublishVolume(ctx context.Context, req *csi.NodePublis
+	// before storing: sort mountFlags slice and join to become single string, as this is what we use to compare
+	askedFlags := req.GetVolumeCapability().GetMount().GetMountFlags()
+	sort.Strings(askedFlags)
+	joinedAskedFlags := strings.Join(askedFlags[:], ",")
```
Some code duplication, that we can avoid by reusing the already prepared mount flags array here: https://github.com/intel/pmem-csi/pull/463/commits/fd89169bb0b4c245157c40d554035d0a5353bae8#diff-4a0caa0c2e37db715640b7cbf97d7a1fR212-R214

okartau replied:
The code above is inside if-block so it is not guaranteed it has created that value.

okartau replied:
what about we move that computation above the "if" so that it happens always once?

avalluri replied:
> what about we move that computation above the "if" so that it happens always once?

Yes I think, as we always store that info anyway.

avalluri commented on pkg/pmem-csi-driver/nodeserver.go:
```diff
@@ -374,6 +413,7 @@ func (ns *nodeServer) NodeUnstageVolume(ctx context.Context, req *csi.NodeUnstag
 		return nil, err
 	}
 
+	ns.volFstype[req.VolumeId] = ""
```
When exactly we clean/remove the cache cahced map(s) elements?

okartau replied:
thanks, indeed I should use delete() instead of just overwriting with empty, added delete() calls in latest code.

avalluri commented on pkg/pmem-csi-driver/nodeserver.go:
```diff
@@ -199,14 +199,12 @@ func (ns *nodeServer) NodePublishVolume(ctx context.Context, req *csi.NodePublis
-			//   Means, we check only mountflags part of Capability.
-			//   Mount flags are coded as []string, so we sort and join mount flags to single string
-			//   for easier comparing.
+			//    We do not check VolumeCapability/fsType as fsType may be missing (is used in NodeStageVolume).
```
These remvoed comments are added part of this PR, I feel like you can squash with your earlier commit which introduced these comments. I am ok to completely sqashing this commit too.

okartau replied:
thanks, yes I had changed in later commit that I had created in previous one. I now combined the changes of text-changing commit into previous bigger commit.

okartau commented:
addressed recent commits, force-pushed, also e2e testing that code locally while CI is not yet functioning.

pohly commented:
We have good CI results for this one. @avalluri, is it okay to merge?

okartau commented:
I pushed a version where mountflags sorting for storing is made above "if" to have single instance of that code. 
This PR needs more work to handle fsType values storing, based on off-line discussion we had with @avalluri, thus restoring WIP in title

okartau commented:
pushed with added code to compare also fsType, but only if present in request.
But this now happens only in NodePublish. I wonder should we also add fsType check request vs existing in NodeStage.

okartau commented:
pushed new version where main improvement commit is changed so that it get fsType from device instead of storing in map.

okartau commented:
pushed a new version again, checking of existing file system is moved out from ProvisionDevice.

okartau commented:
One more thing, let's combine the change here once more with #453 and see the CI result. It used to be combined, then I removed #453 part, but then we changed it some more, so the latest state or PR change is not verified combined with repeated tests.

okartau commented:
As this CI failure here demonstrates (and also failure on #453 after rebase), recent change to strictly forcing vendor/ state takes away method that I use for development purposes: having local changes in devel/, but only in PR, not to be merged. I combine such changes with other PRs to generate a state with those changes that are not (yet) upstreamed. Is this how we want it, (i.e. we cant put changes through CI with intentional vendor/ changes that we know of but we plan not to check in), or could we have some method to work around that check?
I understand this protects us against unintentional vendor/ changes, but what about intentional.

pohly commented:
Olev Kartau <notifications@github.com> writes:
> I understand this protects us against unintentional vendor/ changes,
> but what about intentional.

Remove the "make test" step in the Jenkinsfile?

okartau commented:
> Remove the "make test" step in the Jenkinsfile?

Tried that in #453 but it fails in dm-test. Does dm-test depend on something that happens in 'make test' ? It's not that obvious from output what went wrong, or is it for experienced eye...
https://cloudnative-k8sci.southcentralus.cloudapp.azure.com/blue/organizations/jenkins/pmem-csi/detail/PR-453/4/pipeline

avalluri commented:
```
[2019-11-25T15:05:08.372Z] + make run_dm_tests
[2019-11-25T15:05:08.372Z] curl -L https://github.com/govm-project/govm/releases/download/0.9-alpha/govm_0.9-alpha_Linux_x86_64.tar.gz -o /mnt/workspace/pmem-csi_PR-453/_work/govm_0.9-alpha_Linux_amd64.tar.gz
[2019-11-25T15:05:08.372Z]   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
[2019-11-25T15:05:08.372Z]                                  Dload  Upload   Total   Spent    Left  Speed
[2019-11-25T15:05:08.372Z] 
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100   625    0   625    0     0   2192      0 --:--:-- --:--:-- --:--:--  2192
[2019-11-25T15:05:08.372Z] Warning: Failed to create the file 
[2019-11-25T15:05:08.372Z] Warning: /mnt/workspace/pmem-csi_PR-453/_work/govm_0.9-alpha_Linux_amd64.tar.gz
[2019-11-25T15:05:08.372Z] Warning: : No such file or directory
```
faile at the `govm `installation step.

okartau commented:
you mean failure was independent of skip of 'make test' step ? and it could work when retried?

avalluri commented:
> you mean failure was independent of skip of 'make test' step ? and it could work when retried?

It should but, `_work/govm_$(GOVM_VERSION)_Linux_amd64.tar.gz:` target is not creating `_work` dir. This was not visible as `make test`(maybe `setup-ca.sh`) crates that.

pohly commented:
Olev Kartau <notifications@github.com> writes:
> Tried that in #453 but it fails in dm-test. Does dm-test depend on
> something that happens in 'make test' ?

Let's just get rid of it, okay?

See https://github.com/intel/pmem-csi/pull/473

okartau commented:
CI success in combination with repeated tests, removing WIP and repeating-dev-debug commits, force pushed

pohly commented:
Is this ready for merging or does it need another round of review by @avalluri?

avalluri commented:
> Is this ready for merging or does it need another round of review by @avalluri?

I can be merged., but needs a follow-up PR to cover the points mentioned here https://github.com/intel/pmem-csi/pull/463#pullrequestreview-322116750

okartau commented:
> but needs a follow-up PR

I created issue #483, will merge this.
